fred2/customstringsdlg.cpp

```diff
@@ -0,0 +1,375 @@
+	CString text_str;
+	text_edit->GetWindowText(text_str);
+	m_custom_strings[i].text = text_str;
+
+	update_data_lister();
+	update_text_edit_boxes("", "", "");
+}
+
+bool CustomStringsDlg::edit_boxes_have_valid_data()
+{
+	if (!data_edit_box_has_valid_data()) {
+		return false;
+	}
+
+	if (!text_edit_box_has_valid_data()) {
+		return false;
+	}
+
+	if (!key_edit_box_has_valid_data()) {
+		return false;
+	}
+
+	return true;
+}
+
+bool CustomStringsDlg::key_edit_box_has_valid_data()
+{
+
+	CEdit* key_edit = (CEdit*)GetDlgItem(IDC_CUSTOM_KEY);
+	CString key_str;
+	key_edit->GetWindowText(key_str);
+
+	if (key_str.IsEmpty()) {
+		MessageBox("You have not entered a key!");
+		return false;
+	}
+
+	if (key_str.Find(" ") != -1) {
+		MessageBox("Names cannot contain spaces!");
+		return false;
+	}
+
+	// Only check for duplicate keys if there are other keys to check!
+	if (m_lister_keys.size() > 0) {
+		const int index = m_data_lister.GetCurSel();
+		const auto& this_key = m_lister_keys[index];
+
+		if (strcmp(this_key.c_str(), key_str)) {
+			for (const auto& cs : m_custom_strings) {
+				const CString key = cs.name.c_str();
+				if (key == key_str) {
+					MessageBox("Names must be unique!");
+					return false;
+				}
+			}
+		}
+	}
+
+	return true;
+}
+
+bool CustomStringsDlg::data_edit_box_has_valid_data()
+{
+	CEdit* data_edit = (CEdit*)GetDlgItem(IDC_CUSTOM_DATA);
+	CString data_str;
+	data_edit->GetWindowText(data_str);
+
+	if (data_str.IsEmpty()) {
+		MessageBox("You have not entered any data!");
+		return false;
+	}
+
+	return true;
+}
+
+bool CustomStringsDlg::text_edit_box_has_valid_data()
+{
+	CEdit* text_edit = (CEdit*)GetDlgItem(IDC_CUSTOM_STRING);
+	CString text_str;
+	text_edit->GetWindowText(text_str);
+
+	if (text_str.IsEmpty()) {
+		MessageBox("You have not entered any text!");
+		return false;
+	}
+
+	return true;
+}
+
+void CustomStringsDlg::update_data_lister()
+{
+	m_data_lister.ResetContent();
+	m_lister_keys.clear();
+
+	for (const auto& cs : m_custom_strings) {
+		m_lister_keys.emplace_back(cs.name);
+		const SCP_string lister_entry = cs.name + " - " + cs.value;
+		m_data_lister.AddString(lister_entry.c_str());
+	}
+
+	m_data_lister.EnableWindow(m_data_lister.GetCount() > 0);
+}
+
+void CustomStringsDlg::update_text_edit_boxes(const SCP_string& key, const SCP_string& value, const SCP_string& text)
+{
+	
+	CEdit* key_edit = (CEdit*)GetDlgItem(IDC_CUSTOM_KEY);
+	key_edit->SetWindowText(key.c_str());
+
+	CEdit* data_edit = (CEdit*)GetDlgItem(IDC_CUSTOM_DATA);
+	data_edit->SetWindowText(value.c_str());
+
+	CEdit* data_text = (CEdit*)GetDlgItem(IDC_CUSTOM_STRING);
+	data_text->SetWindowText(text.c_str());
+}
+
+void CustomStringsDlg::OnEnSetFocusEditString()
+{
+	m_text_edit_focus = true;
+}
+
+void CustomStringsDlg::OnEnKillFocusEditString()
+{
+	m_text_edit_focus = false;
+}
+
+bool CustomStringsDlg::query_modified() const
+{
+	//return The_mission.custom_strings != m_custom_strings;
```

Comment:
I'm guessing this wasn't working?   Sounds like you would need a nested loop to properly check if anything has been altered.

While most of the code benefits from using a vector, I think this is where using an unordered map (hash map in other languages) would have made things a lot easier here.  Anything where you deal with something having unique keys, an ordered or unordered map will probably have benefits.

Reply:
Right. It took me a while to remember this one... FRED doesn't seem to like comparing vectors or even std::equal. So I've implemented a poor-man's version here now instead. It'll do the job.